{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import sys\n",
    "sys.path.append('./')\n",
    "import auxiliary_bed_functions as abf\n",
    "chr_list = abf.getChrList()\n",
    "\n",
    "RNAseq_input = '../tensor_input/RNAseq_input/'\n",
    "output_dir = '../'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the canonical gene table for the hg19 genome.\n",
    "# The table was downloaded from the UCSC table browser.\n",
    "# Select the knownCanonical table and select the\n",
    "# three genomic coordinate fields, known gene ID\n",
    "# and geneSymbol fields. The canonical gene regions\n",
    "# are used for defining a gene regulatory region to \n",
    "# compare to histone modification data\n",
    "gene_file = open('./knownCanonical_hg19.txt','r')\n",
    "\n",
    "gene_table = []\n",
    "for gene in gene_file:\n",
    "    if(gene[0] != '#'):\n",
    "        if(gene.strip().split('\\t')[0] not in chr_list):\n",
    "            continue\n",
    "        gene_table.append(gene.strip().split('\\t'))\n",
    "        \n",
    "gene_file.close()\n",
    "\n",
    "\n",
    "# The isoform table can be obtained from the UCSC Table Browser.\n",
    "# Select the knownGene table for the hg19 genome and select the\n",
    "# first 5 fields for output.\n",
    "isoform_file = open('./knownGene_hg19.txt','r')\n",
    "\n",
    "isoform_table = []\n",
    "for iso in isoform_file:\n",
    "    if(iso[0] != '#'):\n",
    "        isoform_table.append(iso.strip().split('\\t'))\n",
    "    \n",
    "isoform_file.close()\n",
    "\n",
    "# Get a strand for each gene in the canonical\n",
    "# gene table. If the official protein symbol from the\n",
    "# canonical gene table has multiple isoforms in the \n",
    "# isoform table, the strand of the first isoform is used.\n",
    "# The strand is used to define the TSS as 5' coordinate \n",
    "# on the respected strand.\n",
    "for i in range(0,len(gene_table)):\n",
    "    for iso in isoform_table:\n",
    "        if(gene_table[i][3] == iso[0]):\n",
    "            gene_table[i].append(iso[2])\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the TSS for each canonical gene and sort these by\n",
    "# chromosome.\n",
    "\n",
    "TSS_sorted_genes = [[] for c in chr_list]\n",
    "TSS_loc = [[] for c in chr_list]\n",
    "\n",
    "for i in range(0,len(gene_table)):\n",
    "    gene_table[i][1] = int(gene_table[i][1])\n",
    "    gene_table[i][2] = int(gene_table[i][2])\n",
    "\n",
    "gene_table_byChr = [[] for i in range(0,len(chr_list))]\n",
    "for i in range(len(gene_table)):\n",
    "    gene_table_byChr[chr_list[gene_table[i][0]]].append(gene_table[i]) \n",
    "            \n",
    "for c in range(len(chr_list)):\n",
    "    for i in range(len(gene_table_byChr[c])):\n",
    "        if(gene_table_byChr[c][i][5] == '+'):\n",
    "            TSS_loc[c].append(gene_table_byChr[c][i][1])\n",
    "        elif(gene_table_byChr[c][i][5] == '-'):\n",
    "            TSS_loc[c].append(gene_table_byChr[c][i][2])\n",
    "        else:\n",
    "            print('No strand specified', c, i)       \n",
    "            \n",
    "for c in range(len(chr_list)):\n",
    "    TSS_sort = np.sort(TSS_loc[c])\n",
    "    TSS_index = np.argsort(TSS_loc[c])\n",
    "    for i in range(len(TSS_sort)):\n",
    "        TSS_sorted_genes[c].append([gene_table_byChr[c][TSS_index[i]][0],\n",
    "                                    TSS_sort[i], gene_table_byChr[c][TSS_index[i]][3],\n",
    "                                    gene_table_byChr[c][TSS_index[i]][4]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each TSS get the region extending 10kb in the 5' and 3'\n",
    "# direction of the TSS.\n",
    "reg_regions = open('../gene_TSS_regulatory_regions10kb.tsv','w+')\n",
    "for c in range(len(chr_list)):\n",
    "    for i in range(len(TSS_sorted_genes[c])):\n",
    "        reg_regions.write(\"{}\\t{}\\t{}\\t{}\\t{}\\n\".format(TSS_sorted_genes[c][i][0],\n",
    "                                                               max(int(TSS_sorted_genes[c][i][1])-10000,0),\n",
    "                                                               int(TSS_sorted_genes[c][i][1])+10000,\n",
    "                                                               TSS_sorted_genes[c][i][2],\n",
    "                                                               TSS_sorted_genes[c][i][3]))\n",
    "reg_regions.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Filter genes by FPKM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The hg19 knownGene gtf annotation file. This\n",
    "# is used to obtain the gene length for FPKM\n",
    "# calculations.\n",
    "gtf_file_name = './genes_hg19.gtf'\n",
    "\n",
    "# Feature counts for each of the samples\n",
    "file_list_L = ['L_pt1_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt2_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt3_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt4_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt5_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt6_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt7_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt8_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt9_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt10_1_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt11_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt12_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt13_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt14_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt15_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt17_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt18_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt19_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt20_RNAseq_alignReadsPerGene.tsv',\n",
    "               'L_pt21_RNAseq_alignReadsPerGene.tsv']\n",
    "\n",
    "file_list_M = ['M_pt1_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt2_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt3_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt4_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt5_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt6_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt7_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt8_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt9_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt10_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt11_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt12_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt13_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt14_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt15_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt17_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt18_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt19_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt20_RNAseq_alignReadsPerGene.tsv',\n",
    "               'M_pt21_RNAseq_alignReadsPerGene.tsv']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def lengthOfUnion(union):\n",
    "    # Get the length of the union of exons on each \n",
    "    # chromosome. Return the minimal length if exons\n",
    "    # are found on multiple chromosomes for a given\n",
    "    # gene_id.\n",
    "    length = [0 for c in range(len(chr_list))]\n",
    "    for c, chrom in enumerate(chr_list):\n",
    "        for i in range(len(union[c])):\n",
    "            length[c] += union[c][i][2]-union[c][i][1] + 1\n",
    "    return min(np.array(length)[np.where(np.array(length) > 0)[0]])\n",
    "\n",
    "def lengthOfExons(region_list):\n",
    "    # Take the union of the exons on each chromosome.\n",
    "    region_chr_list = [[] for c in range(len(chr_list))]\n",
    "    for region in region_list:\n",
    "        region_chr_list[chr_list[region[0]]].append(region)\n",
    "    get_union = abf.unionOfRegions(region_chr_list, 1)\n",
    "    return lengthOfUnion(get_union)\n",
    "\n",
    "def readGTF(gtf_file_name):\n",
    "    # Read the gtf file and get the total exon length\n",
    "    # for each gene_id.\n",
    "    gtf_file = open(gtf_file_name, 'r')\n",
    "    gene_list = []\n",
    "    gene_dic = {}\n",
    "    n = 0\n",
    "    region_list = []\n",
    "    for line in gtf_file:\n",
    "        l = line.strip().split('\\t')\n",
    "        if(l[2] != 'exon'):\n",
    "            continue\n",
    "        if(l[0] not in chr_list):\n",
    "            continue\n",
    "        # Extract the gene_id\n",
    "        gene_name = l[-1].split(';')[0]\n",
    "        gene_name = gene_name.split(' ')[1]\n",
    "        gene_name = gene_name.strip('\"') \n",
    "        gene_list.append(gene_name)\n",
    "        # Get the list of all exons for each gene_id.\n",
    "        if(gene_name not in gene_dic):\n",
    "            gene_dic[gene_name] = [[l[0],int(l[3]),int(l[4])]]\n",
    "        else:\n",
    "            gene_dic[gene_name].append([l[0],int(l[3]),int(l[4])])\n",
    "    # Get the length of the union of all exons.\n",
    "    gene_length_list = {}\n",
    "    for gene in gene_dic:\n",
    "        length = lengthOfExons(gene_dic[gene])\n",
    "        gene_length_list[gene] = length\n",
    "    return gene_length_list\n",
    "\n",
    "def getFPKM(gene_count_list, gene_length):\n",
    "    # Given the file of gene counts (two coloumns, one \n",
    "    # for gene name and one for the number of reads aligning\n",
    "    # to this gene) and a dictionary of gene lengths, compute\n",
    "    # the FPKM for each gene.\n",
    "    gene_FPKM = [[] for i in range(len(gene_count_list))]\n",
    "    total = sum([gene_count_list[i][1] for i in range(len(gene_count_list))])\n",
    "    for i in range(len(gene_count_list)):\n",
    "        FPKM = (float(gene_count_list[i][1]) * 1e9)/(float(gene_length[gene_count_list[i][0]])*total)\n",
    "        gene_FPKM[i] = [gene_count_list[i][0], FPKM]\n",
    "\n",
    "    return gene_FPKM\n",
    "\n",
    "def readGeneCount(gene_count_name):\n",
    "    # Read the gene count files.\n",
    "    gene_count_file = open(gene_count_name)\n",
    "    gene_count_list = []\n",
    "    for line in gene_count_file:\n",
    "        l = line.strip().split('\\t')\n",
    "        gene_count_list.append([l[0],int(l[1])])    \n",
    "    return gene_count_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_length = readGTF(gtf_file_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "FPKM_matrix = np.zeros((26334, 2*len(file_list_L)))\n",
    "for i in range(len(file_list_M)):\n",
    "    FPKM_list = getFPKM(readGeneCount(RNAseq_input + file_list_M[i]),gene_length)\n",
    "    for j in range(len(FPKM_list)):\n",
    "        FPKM_matrix[j][i] = FPKM_list[j][1]\n",
    "        \n",
    "for i in range(len(file_list_L)):\n",
    "    FPKM_list = getFPKM(readGeneCount(RNAseq_input + file_list_L[i]),gene_length)\n",
    "    for j in range(len(FPKM_list)):\n",
    "        FPKM_matrix[j][i+len(file_list_M)] = FPKM_list[j][1]\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "read_counts = readGeneCount(RNAseq_input + file_list_L[0])\n",
    "gene_names = [read_counts[i][0] for i in range(len(read_counts))]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select only genes with at least 1 FPKM in at\n",
    "# least one of the samples.\n",
    "FPKM_gene_list = []\n",
    "for i in range(len(FPKM_matrix)):\n",
    "    m = 0\n",
    "    for j in range(len(FPKM_matrix[i])):\n",
    "        if(FPKM_matrix[i][j] >= 1.0):\n",
    "            m+=1\n",
    "    if(m >= 1):\n",
    "        FPKM_gene_list.append(gene_names[i])\n",
    "\n",
    "# Write the list of expressed genes for later use\n",
    "output = open(output_dir + 'expressedGenes_final_any.txt','w+')\n",
    "for i in range(len(FPKM_gene_list)):\n",
    "    # Remove genes with length less than 200\n",
    "    if(gene_length[FPKM_gene_list[i]] <= 200):\n",
    "        continue\n",
    "    output.write(\"{}\\n\".format(FPKM_gene_list[i]))\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
